pkg/template/resource.go

```diff
@@ -46,40 +47,58 @@ type getClusterTriggerBinding func(name string, options metav1.GetOptions) (*tri
 // ResolveTrigger takes in a trigger containing object refs to bindings and
 // templates and resolves them to their underlying values.
 func ResolveTrigger(trigger triggersv1.EventListenerTrigger, getTB getTriggerBinding, getCTB getClusterTriggerBinding, getTT getTriggerTemplate) (ResolvedTrigger, error) {
-	tb := make([]*triggersv1.TriggerBinding, 0, len(trigger.Bindings))
-	ctb := make([]*triggersv1.ClusterTriggerBinding, 0, len(trigger.Bindings))
-	for _, b := range trigger.Bindings {
-		if b.Spec != nil {
-			tb = append(tb, &triggersv1.TriggerBinding{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: b.Name,
-				},
-				Spec: *b.Spec,
+	bp, err := resolveBindingsToParams(trigger.Bindings, getTB, getCTB)
+	if err != nil {
+		return ResolvedTrigger{}, fmt.Errorf("failed to resolve bindings: %w", err)
+	}
+	tt, err := getTT(trigger.Template.Name, metav1.GetOptions{})
+	if err != nil {
+		return ResolvedTrigger{}, fmt.Errorf("error getting TriggerTemplate %s: %w", trigger.Template.Name, err)
+	}
+	return ResolvedTrigger{TriggerTemplate: tt, BindingParams: bp}, nil
+}
+
+// resolveBindingsToParams takes in both embedded bindings and references and returns a list of resolved Param values.ResolveBindingsToParams
+func resolveBindingsToParams(bindings []*triggersv1.TriggerSpecBinding, getTB getTriggerBinding, getCTB getClusterTriggerBinding) ([]triggersv1.Param, error) {
+	bindingParams := []triggersv1.Param{}
+	for _, b := range bindings {
+		switch {
+		case b.Spec != nil: // Could also call SetDefaults and not rely on this?
+			bindingParams = append(bindingParams, b.Spec.Params...)
```

Comment:
Won't it cause duplication error (https://github.com/tektoncd/triggers/pull/768/files#diff-7b6fd7ee11e66c6319df083beac6f30aR97) if SetDefaults has already been called?

Reply:
ahh good point.

Reply:
actually, looking through the code, I'm not sure it will. In SetDefaults, we convert the Spec.Params to name/value pairs. We do not keep the Spec field after.

The case here will only apply for cases where set defaults has not been called i.e. if someone upgrades to this version with an existing EL that contains an embedded binding.